server/datastore/mysql/operating_systems.go

```diff
@@ -0,0 +1,144 @@
+package mysql
+
+import (
+	"context"
+	"database/sql"
+	"errors"
+
+	"github.com/fleetdm/fleet/v4/server/contexts/ctxerr"
+	"github.com/fleetdm/fleet/v4/server/fleet"
+	"github.com/jmoiron/sqlx"
+)
+
+func (ds *Datastore) ListOperatingSystems(ctx context.Context) ([]fleet.OperatingSystem, error) {
+	return listOperatingSystemsDB(ctx, ds.reader)
+}
+
+func listOperatingSystemsDB(ctx context.Context, tx sqlx.QueryerContext) ([]fleet.OperatingSystem, error) {
+	var os []fleet.OperatingSystem
+	if err := sqlx.SelectContext(ctx, tx, &os, `SELECT * FROM operating_systems`); err != nil {
+		return nil, err
+	}
+	return os, nil
+}
+
+func (ds *Datastore) UpdateHostOperatingSystem(ctx context.Context, hostID uint, hostOS fleet.OperatingSystem) error {
+	os, err := maybeNewOperatingSystemDB(ctx, ds.writer, hostOS)
+	if err != nil {
+		return err
+	}
+	return maybeUpdateHostOperatingSystemDB(ctx, ds.writer, hostID, os.ID)
+}
+
+// maybeNewOperatingSystemDB queries the `operating_systems` table with the
+// name, version, arch, and kernel_version of the given operating system. If found,
+// it returns the record including the associated ID. If not found, it returns a call
+// to `newOperatingSystemDB`, which inserts a new record and returns the record
+// including the newly associated ID.
+func maybeNewOperatingSystemDB(ctx context.Context, tx sqlx.ExtContext, hostOS fleet.OperatingSystem) (*fleet.OperatingSystem, error) {
+	switch os, err := getOperatingSystemDB(ctx, tx, hostOS); {
+	case err == nil:
+		return os, nil
+	case errors.Is(err, sql.ErrNoRows):
+		return newOperatingSystemDB(ctx, tx, hostOS)
```

Comment:
Don't you need to make a call to `insertHostOperatingSystemDB` here as well?

Reply:
The function just handles the `operating_systems` table. The `host_operating_system` table insert/update is a separate function. It gets called at line 30 above.